### PR TITLE
android 12 fix

### PR DESCRIPTION
**Pull request by FilippiMatteo**
Please merge this small changes for android 12 with build-tools sdk 31.
thank you

### CODE REVIEW

File: plugin.xml [hunk 1]
```diff
@@ -72,7 +72,7 @@
 
 
     <config-file target="AndroidManifest.xml" parent="/*/application">
-      <receiver android:name="nl.xservices.plugins.ShareChooserPendingIntent" android:enabled="true">
+      <receiver android:name="nl.xservices.plugins.ShareChooserPendingIntent" android:enabled="true" android:export="true">
```
Comment (vouillon) [inline plugin.xml:75]:
```suggestion
      <receiver android:name="nl.xservices.plugins.ShareChooserPendingIntent" android:enabled="true" android:export="true">
```
```suggestion
      <receiver android:name="nl.xservices.plugins.ShareChooserPendingIntent" android:enabled="true" android:exported="true">
```

File: src/android/nl/xservices/plugins/SocialSharing.java [hunk 2]
```diff
@@ -271,8 +271,7 @@ public void run() {
         final boolean hasMultipleAttachments = files.length() > 1;
         final Intent sendIntent = new Intent(hasMultipleAttachments ? Intent.ACTION_SEND_MULTIPLE : Intent.ACTION_SEND);
         final Intent receiverIntent = new Intent(cordova.getActivity().getApplicationContext(), ShareChooserPendingIntent.class);
-        final PendingIntent pendingIntent = PendingIntent.getBroadcast(cordova.getActivity().getApplicationContext(), 0, receiverIntent, PendingIntent.FLAG_UPDATE_CURRENT);
-        sendIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
+final PendingIntent pendingIntent = PendingIntent.getBroadcast(cordova.getActivity().getApplicationContext(), 0, receiverIntent, PendingIntent.FLAG_UPDATE_CURRENT|PendingIntent.FLAG_IMMUTABLE);        sendIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
```
Comment (vouillon) [inline src/android/nl/xservices/plugins/SocialSharing.java:274]:
```suggestion
        final PendingIntent pendingIntent = PendingIntent.getBroadcast(cordova.getActivity().getApplicationContext(), 0, receiverIntent, PendingIntent.FLAG_UPDATE_CURRENT|PendingIntent.FLAG_IMMUTABLE);
        sendIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
```

**Comment by bhandaribhumin** [comment]
@FilippiMatteo @vouillon you need to add `android:exported="true/false"` at [nl.xservices.plugins.ShareChooserPendingIntent](https://github.com/EddyVerbruggen/SocialSharing-PhoneGap-Plugin/blob/master/plugin.xml#L75)

Right now i was handling by adding below code in config.xml

```
 <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application/activity[@android:name='nl.xservices.plugins.ShareChooserPendingIntent']">
            <activity android:exported="true" />
  </edit-config>
```

**Comment by bhandaribhumin** [comment]
any plan for merge to master?

**Comment by jmalfatto** [comment]
> any plan for merge to master?

I second this question!

**Comment by ucsbricks** [comment]
Thanks @FilippiMatteo 

@EddyVerbruggen any plans to merge his PR?

**Comment by tomfong** [comment]
@EddyVerbruggen Please merge this PR asap.

**Comment by hanslbua** [comment]
does anyone know, when this will be merged? i need this feature to support android 12

**Comment by tomfong** [comment]
> does anyone know, when this will be merged? i need this feature to support android 12

no one knows... I've already tweeted Eddy but not yet got the reply. The temporary solution is to manually change the code according to this PR every time before building the app in Android Studio, if you use capacitor.

**Comment by hldh214** [comment]
> @FilippiMatteo @vouillon you need to add `android:exported="true/false"` at [nl.xservices.plugins.ShareChooserPendingIntent](https://github.com/EddyVerbruggen/SocialSharing-PhoneGap-Plugin/blob/master/plugin.xml#L75)
> 
> Right now i was handling by adding below code in config.xml
> 
> ```
>  <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application/receiver[@android:name='nl.xservices.plugins.ShareChooserPendingIntent']">
>             <activity android:exported="true" />
>   </edit-config>
> ```

May I ask which section should this xml stay?
I tried to put it under `<widget>` in `config.xml` and when I remove `platform/android` folder and do `cordova platform add android`, I got error says this `ShareChooserPendingIntent` section not found(see below), I guess the plugin's discovery process is after this edit config process, so how do we deal with this situation?

```
Subproject Path: CordovaLib
Subproject Path: app
Android project created with cordova-android@9.1.0
Unable to graft xml at selector "/manifest/application/receiver[@android:name='nl.xservices.plugins.ShareChooserPendingIntent']" from "/app/platforms/android/app/src/main/AndroidManifest.xml" during config install
```

**Comment by EinfachHans** [comment]
Please merge this!

**Review by tomfong — APPROVED** [review]
@EddyVerbruggen Please approve and merge this PR.

**Comment by wata-tomnguyen** [comment]
Hi @EddyVerbruggen
Please merge this PR, it's causing a stuck in my project when i updated it to android 12
Thanks bro!

**Comment by whria78** [comment]
This solution works for me.

**Comment by Takkuz** [comment]
Due to the lack of PR merge I forked it and published on NPM as `cordova-plugin-x-socialsharing-android12` with the commits done by @FilippiMatteo as v6.0.4.

You can use it by installing as `npm i cordova-plugin-x-socialsharing-android12` and `cordova plugin add cordova-plugin-x-socialsharing-android12`

I think I'll remove it as soon as this will be merged.

**Comment by rohanpatil-pfizer** [comment]
Is it merged now ?

**Comment by tomfong** [comment]
@Takkuz How about for Capacitor? How to integrate yours to `awesome-cordova-plugin`?

**Comment by Takkuz** [comment]
> @Takkuz How about for Capacitor? How to integrate yours to `awesome-cordova-plugin`?

Sorry, I'm not used to Capacitor so I don't know how to integrate it.

**Comment by bonjourjoel** [comment]
Hi, i have the same problem.

**Comment by HarelM** [comment]
@Takkuz thanks for this! Any chance you can remove the dependency for `es6-promise-plugin` in your npm version?
@tomfong there's no issue with using @Takkuz plugin with capacitor - npm uninstall "old-one" && npm i "new-one" && npx cap sync, as usual
You can use it with `@ionic-native/social-sharing` so that it is an angular injectable serivce.

**Comment by TomCosta** [comment]
> Due to the lack of PR merge I forked it and published on NPM as `cordova-plugin-x-socialsharing-android12` with the commits done by @FilippiMatteo as v6.0.4.
> 
> You can use it by installing as `npm i cordova-plugin-x-socialsharing-android12` and `cordova plugin add cordova-plugin-x-socialsharing-android12`
> 
> I think I'll remove it as soon as this will be merged.

It works great, thanks. But if someone else like a team member clones the code, it doesn't install the same plugin again. It install the older version of the plugin instead of the new one. 
Please, how may we solve this?

**Comment by Takkuz** [comment]
> > Due to the lack of PR merge I forked it and published on NPM as `cordova-plugin-x-socialsharing-android12` with the commits done by @FilippiMatteo as v6.0.4.
> > You can use it by installing as `npm i cordova-plugin-x-socialsharing-android12` and `cordova plugin add cordova-plugin-x-socialsharing-android12`
> > I think I'll remove it as soon as this will be merged.
> 
> It works great, thanks. But if someone else like a team member clones the code, it doesn't install the same plugin again. It install the older version of the plugin instead of the new one. Please, how may we solve this?

Ok, I should have fixed it. It should be due to the ald ID still used inside plugin section. I don't have it because I don't use the plugin section on config.xml.

**Comment by HarelM** [comment]
@Takkuz any chance to remove the es6 plugin dependency while your are at it?

**Comment by Takkuz** [comment]
> @Takkuz any chance to remove the es6 plugin dependency while your are at it?

I was looking at it but i don't understand native code, so I don't know why it's needed and if it could be removed without affecting the behaviour. Actually I'm on holiday and my company needs a working production of this plugin so I can't take the risk. Sorry.

**Comment by HarelM** [comment]
I'm using your plugin without this dependency and it works as expected. I don't think it's needed anymore as all modern browsers supports promises.

**Comment by dcxn** [comment]
Have the same issue. Guess no one can move this along other than @EddyVerbruggen 😓

I tried the solution frequently suggested here https://github.com/EddyVerbruggen/SocialSharing-PhoneGap-Plugin/pull/1158#issuecomment-1103770496 but that creates the `Unable to graft xml at selector` error @hldh214  mentioned.

**Comment by EltonFaust** [comment]
The `android:exported="true"` problem on the main activity, was fixed in the newer versions of the android-cordova ([check here in release notes](https://github.com/apache/cordova-android/blob/60e3803c67734087b7a6ff8ee8bd3a659cb050f8/RELEASENOTES.md#1012-apr-11-2022)).
Install the newest version will prevent the problem, preventing the need of the `edit-config` option.
You can install de newest version with:
```sh
cordova platform remove android
cordova platform add android@11
```
_Be aware that the newer versions use the java 11 instead of the java 8._

@FilippiMatteo I [opened a pull](https://github.com/FilippiMatteo/SocialSharing-PhoneGap-Plugin/pull/1) on your branch, it's just a fix, since the min SDK version allowed on cordova-android is 22, and the flag `FLAG_IMMUTABLE` [was added only on SKD 23](https://developer.android.com/reference/android/app/PendingIntent#FLAG_IMMUTABLE), to prevent some unexpected behaviour on devices with SDK 22, I only added the flag when needed, if you could check the pull and merge into yours.

**Comment by bonjourjoel** [comment]
> The `android:exported="true"` problem on the main activity, was fixed in the newer versions of the android-cordova ([check here in release notes](https://github.com/apache/cordova-android/blob/60e3803c67734087b7a6ff8ee8bd3a659cb050f8/RELEASENOTES.md#1012-apr-11-2022)). Install the newest version will prevent the problem, preventing the need of the `edit-config` option. You can install de newest version with:
> 
> ```shell
> cordova platform remove android
> cordova platform add android@11
> ```

Hi. I have the latest cordova android, installed like you mentioned. However I do have the problem with android:exported="true".
I had to add an edit-config, plus some weird fix in my prepare hook to make the packaging work.
And I still need a fix on this plugin.

And all this is related to the latest cordova android release AND this plugin not being compatible.

Regards.

**Comment by EltonFaust** [comment]
Weird it didn't work on the main activity with the newer version (the newer version don't fix the plugin problem only on the main activity ploblem), i've had a lot of problems adding/removing plugins and the platform itself, I would recommend to remove entirely the node_modules, plugins and platforms directories and try add the platform again, solved a lot of problems with that...

**Comment by bonjourjoel** [comment]
> Weird it didn't work on the main activity with the newer version (the newer version don't fix the plugin problem only on the main activity ploblem), i've had a lot of problems adding/removing plugins and the platform itself, I would recommend to remove entirely the node_modules, plugins and platforms directories and try add the platform again, solved a lot of problems with that...

The problem is not on the main activity. The problem is on the activity of the plugin.
The main activity is fine and doesn't produce an error.

**Comment by EltonFaust** [comment]
My mistake, I thought was understood that only the `android:exported="true"` wont be enough, this will allow the app to be built, but once the plugin is called, and reaches the `PendingIntent` without the mutability flag, will crash your app (on devices with SDK 31+).

Since someone provided an alternative npm package with all the changes, and seen the log on one comment showing that was used the cordova-android 9, I assumed that the problem still happening only could be on the main activity not on this plugin receiver.

**Comment by luvenan** [comment]
Since the last comments on this are very recent... I guess no updates. This is really bad, because now googleplay requires you to target sdk 31, so this needs to work for the app to be uploaded.

**Comment by HarelM** [comment]
Android 12 package mentioned here is working as expected and can be used. I have it in production now.

**Review by mayconn-netpark — APPROVED** [review]

**Review by dcxn — APPROVED** [review]

**Comment by saggio89** [comment]
This fix worked for me. I will also tweet him to merge it

**Comment by HarelM** [comment]
@EddyVerbruggen can you remove the es6-promise dependency?

**Comment by tomfong** [comment]
finally merged!

**Comment by bonjourjoel** [comment]
Awesome news!

**Comment by idrisAd** [comment]
Hi ! 
As @hldh214 I have the problem below
`Android project created with cordova-android@11.0.0
Unable to graft xml at selector "/manifest/application/receiver[@android:name='nl.xservices.plugins.ShareChooserPendingIntent']" from "/Users/vagrant/git/platforms/android/app/src/main/AndroidManifest.xml" during config install`

I added the differents <edit-config> for activity and receiver in my config.xml
`<edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application/activity[@android:name='MainActivity']" xmlns:android="http://schemas.android.com/apk/res/android">
            <activity android:exported="true" />
        </edit-config>
        <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application/receiver[@android:name='nl.xservices.plugins.ShareChooserPendingIntent']">
            <activity android:exported="false" />
        </edit-config>`
I also add cordova-plugin-x-socialsharing but the problem is still here.

Do you have an idea for a solution?

**Comment by saggio89** [comment]
@idrisAd with version 6.0.4 you should be able to delete those "edit-config" lines you added and it should work without them. Also when upping your target SDK to 31 for android make sure you have Java 11 installed. I  had Java 1.8 and it was causing issues building for me.

**Comment by idrisAd** [comment]
Thx @saggio89 ! 
I got errors like `error: cannot find symbol
import android.support.v4.app.NotificationCompat;` but with `cordova plugin add cordova-plugin-androidx-adapter` I solved them.

**Pull request merged**